### PR TITLE
AGX Wireless bug fix

### PR DESCRIPTION
**Pull request by emrahbillur**
Separation of netvmExtraModules caused a bug in agx wireless, where this was fixed by adding appropriate corrections. Wireless condition for agx is improved. A small fix was added for docker (where it is not included if it is not enabled) in case it is planned to add for support.

### CODE REVIEW

**Comment by emrahbillur** [comment]
https://ssrc.atlassian.net/browse/SP-2906 Bug was introduced by netvmExtraModules in hardware was overriding the netvmExtraModules in target. Fixed that issue. There were unnecessary if statements in som type checking for enabling wireless, fixed. Added a fix for docker if somehow being added but this will not introduce any delay as if docker support is not included the fix is also not included.

File: targets/nvidia-jetson-orin.nix
Comment (vilvo) [inline targets/nvidia-jetson-orin.nix:49]: commenting out is probably not intentional?

File: targets/nvidia-jetson-orin.nix
Comment (vilvo) [inline targets/nvidia-jetson-orin.nix:55]: Should we also remove this https://github.com/tiiuae/ghaf/pull/259/files#diff-5e647cec450851446913497aec8e4d19c2440a05aea244e642b0048269d51fbeL106 as redundant in respect to above line?

**Comment by vilvo** [comment]
> Added a fix for docker if somehow being added but this will not introduce any delay as if docker support is not included the fix is also not included.

I'd prefer the WiFi fix and Docker fix as separate commits/PRs as they are not related to each other. There's practical outcome from this - if the Docker overlay was a separate commit, it's easier to revert the overlay commit when we eventually get the fix from the upstream. The whole docker daemon enabling and the docker group are also separate from overlaying the Docker.

**Comment by vilvo** [comment]
`nix build .#packages.aarch64-linux.nvidia-jetson-orin-agx-debug` (using a remote builder) fails to build - it gets stuck to (with `--debug`):
`[2/0/234 built] building ell-0.5.6 on ssh://vilvo@192.168.1.149: debug1: pledge: fork`

`pledge: fork` seems to come from ssh clientloop.c - http://bxr.su/OpenBSD/usr.bin/ssh/clientloop.c#960

I already rebooted the remote builder, but it repeats. I'll check if it's build host related.

**Comment by vilvo** [comment]
> I'll check if it's build host related.
Nope, still the same. Ends up stuck in `pledge: fork` immediately.

**Comment by emrahbillur** [comment]
> I'll seperate both sides and recommit with only WiFi fix.

**Comment by emrahbillur** [comment]
> > I'll check if it's build host related.
> 
> Nope, still the same. Ends up stuck in `pledge: fork` immediately.

I checked the builds for many times for both flash-script and image flakes of agx and image flakes of nx of but I'll try a rebuild switch to check again.

File: targets/nvidia-jetson-orin.nix
Comment (emrahbillur) [reply targets/nvidia-jetson-orin.nix:49]: It will be removed.

**Comment by emrahbillur** [comment]
And a small note to add. The issue in docker was due to a bug in Go compiler. So that fix will also solve issues in many other Go compiler based applications. It is a docker fix but I'll rename the commit of docker fix as "Docker fix due to Go issue".

**Comment by vilvo** [comment]
> > > I'll check if it's build host related.
> > 
> > 
> > Nope, still the same. Ends up stuck in `pledge: fork` immediately.
> 
> I checked the builds for many times for both flash-script and image flakes of agx and image flakes of nx of but I'll try a rebuild switch to check again.

Ok, thanks for confirming. I'm building it now directly on AGX without remote ssh builder connection.

**Comment by emrahbillur** [comment]
> > > > I'll check if it's build host related.
> > > 
> > > 
> > > Nope, still the same. Ends up stuck in `pledge: fork` immediately.
> > 
> > 
> > I checked the builds for many times for both flash-script and image flakes of agx and image flakes of nx of but I'll try a rebuild switch to check again.
> 
> Ok, thanks for confirming. I'm building it now directly on AGX without remote ssh builder connection.

I've removed the docker fix from my local. After confirming build is correct I'll commit and rename the pull request for Wifi Fix. Then I'll later add a PR with Docker fix due to Go compiler issue.

**Comment by vilvo** [comment]
>  I'm building it now directly on AGX without remote ssh builder connection.

Gets stuck on:
```
[nix-shell:~/ghaf]$ nix build .#packages.aarch64-linux.nvidia-jetson-orin-agx-debug
[0/235 built, 0.0 MiB DL] waiting for lock on '/nix/store/28lpb19hzlw9v5a0alckln2dgfr03wmk-ell-0.56','/n`
```
This happens also on local build on `aarch64` hardware.

**Comment by vilvo** [comment]
> Gets stuck on:
> 
> ```
> [nix-shell:~/ghaf]$ nix build .#packages.aarch64-linux.nvidia-jetson-orin-agx-debug
> [0/235 built, 0.0 MiB DL] waiting for lock on '/nix/store/28lpb19hzlw9v5a0alckln2dgfr03wmk-ell-0.56','/n`
> ```
Apparently I'm facing this https://github.com/NixOS/nix/issues/2029

**Comment by emrahbillur** [comment]
> > I'll commit a Wireless fix only PR just after check. So you could try on that.
> 
> Still the same:
> 
> ```
> [nix-shell:~/ghaf]$ git log --oneline -1
> 54bca8e (HEAD, emrah/main) AGX Wireless bug fix
> 
> [nix-shell:~/ghaf]$ nix build .#packages.aarch64-linux.nvidia-jetson-orin-agx-debug
> [0/235 built] waiting for lock on ...
> ```



    I'll commit a Wireless fix only PR just after check. So you could try on that.

Still the same:

[nix-shell:~/ghaf]$ git log --oneline -1
54bca8e (HEAD, emrah/main) AGX Wireless bug fix

[nix-shell:~/ghaf]$ nix build .#packages.aarch64-linux.nvidia-jetson-orin-agx-debug
[0/235 built] waiting for lock on ...



    I'll commit a Wireless fix only PR just after check. So you could try on that.

Still the same:

[nix-shell:~/ghaf]$ git log --oneline -1
54bca8e (HEAD, emrah/main) AGX Wireless bug fix

[nix-shell:~/ghaf]$ nix build .#packages.aarch64-linux.nvidia-jetson-orin-agx-debug
[0/235 built] waiting for lock on ...



    I'll commit a Wireless fix only PR just after check. So you could try on that.

Still the same:

[nix-shell:~/ghaf]$ git log --oneline -1
54bca8e (HEAD, emrah/main) AGX Wireless bug fix

[nix-shell:~/ghaf]$ nix build .#packages.aarch64-linux.nvidia-jetson-orin-agx-debug
[0/235 built] waiting for lock on ...

root@emrah-ThinkPad-P14s-Gen-3:/home/emrah/ghaf-support-for-nx# nix build .#packages.aarch64-linux.nvidia-jetson-orin-nx-debug
warning: Git tree '/home/emrah/ghaf-support-for-nx' is dirty
Then builds with no issues. Same for nix build .#packages.aarch64-linux.nvidia-jetson-orin-agx-debug

**Comment by vilvo** [comment]
I get following lines to the kernel log - multiple times:
```
[  846.797982] INFO: task test-cipher:15875 blocked for more than 724 seconds.
[  846.798207]       Tainted: G           O      5.10.104 #1-NixOS
[  846.798376] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[  846.798595] task:test-cipher     state:D stack:    0 pid:15875 ppid: 15836 flags:0x00000804
[  846.798599] Call trace:
[  846.798608]  __switch_to+0xc8/0x120
[  846.798615]  __schedule+0x2bc/0x824
[  846.798616]  schedule+0x50/0xd0
[  846.798619]  schedule_timeout+0x160/0x1a0
[  846.798620]  __wait_for_common+0xcc/0x220
[  846.798622]  wait_for_completion+0x34/0x40
[  846.798627]  skcipher_recvmsg+0x314/0x3ec
[  846.798632]  sock_read_iter+0xf0/0x120
[  846.798637]  new_sync_read+0x18c/0x1a0
[  846.798639]  vfs_read+0x134/0x1c0
[  846.798640]  ksys_read+0xec/0x10c
[  846.798641]  __arm64_sys_read+0x28/0x34
[  846.798645]  el0_svc_common.constprop.0+0x80/0x1c4
[  846.798647]  do_el0_svc+0x38/0xac
[  846.798650]  el0_svc+0x1c/0x30
[  846.798652]  el0_sync_handler+0xe0/0x10c
[  846.798654]  el0_sync+0x16c/0x180
```

File: targets/nvidia-jetson-orin.nix
Comment (mikatammi) [reply targets/nvidia-jetson-orin.nix:54]: `virtualization.microvm.netvm.extraModules = netvmExtraModules;` <- this needs to be set in both files. The NixOS module system will then combine the two lists and they are added as a two separate modules into the netvm configuration imports.

**Review by mikatammi — APPROVED** [review]

File: targets/nvidia-jetson-orin.nix
Comment (mikatammi) [reply targets/nvidia-jetson-orin.nix:54]: The separation in general between the targets and the modules which I would like to see, is that the targets would contain more, well, example targets which contain more target-specific information, depending on usecase. Some might want to pass through wifi-card and others might not.

Modules should then aim to provide more general modules for use in other projects/flakes, through flake system. For example one might want to create project for Orin NX and not use any VMs at all, when soneone other might want to run it with all the bells and whistles included, and modules should provide easy way to enable/disable various features.

Currently this separation is not very nicely and clearly happening, but we can leave the refactoring and cleanup to further commits.

**Review by mikatammi — APPROVED** [review]

File: targets/nvidia-jetson-orin.nix
Comment (emrahbillur) [reply targets/nvidia-jetson-orin.nix:54]: It was left as it was required in two different files to make netvmExtraModules section merge.

**Comment by emrahbillur** [comment]
> INFO: task test-cipher:15875 blocked for more than 724 seconds

I'm trying to reproduce the same error but could not manage to get this. I remember in some cross compilation configuration cipher part caused errors but still not sure. I'll also try cross build now.

**Comment by mikatammi** [comment]
Are we ready to merge? I'm just testing this and if everything goes ok I will push the merge button

**Comment by mikatammi** [comment]
In my tests everything go ok, and the NetVM connects to the WLAN network

**Comment by vilvo** [comment]
> > INFO: task test-cipher:15875 blocked for more than 724 seconds
> 
> I'm trying to reproduce the same error but could not manage to get this. I remember in some cross compilation configuration cipher part caused errors but still not sure. I'll also try cross build now.

I changed my remote builder from Orin AGX (ghaf/nixox) to M1 (nixos) and it passed building the problematic package - it's still building the `chromium-unwrapped`.

**Pull request merged**
